### PR TITLE
Improve Grapple + Improve Asteroid Dmg

### PR DESCRIPTION
**Pull request by Unix-Code**
#### Changes
* Grapple now has an extra tug when being pulled against current velocity
* Grapple implementation is now a spring joint
* Asteroids deal damage proportional to collision impact force
* Minor: Procedural Generation now has a bool parameter that determines whether a black hole spawns (`chaseLevel`)

### CODE REVIEW

File: Assets/Scenes/Prototype.unity [hunk 1]
```diff
@@ -4405,6 +4405,11 @@ PrefabInstance:
       propertyPath: m_Name
       value: DirectionIndicatorCanvas
       objectReference: {fileID: 0}
+    - target: {fileID: 4915999520967989114, guid: 35c87d254ce1c034f9e041fa103c0289,
```
Comment (chuck-stein) [inline Assets/Scenes/Prototype.unity:4408]: Did you make any meaningful changes to the scene? If not, can you rollback to Prototype.unity from master, to avoid merge conflicts?

File: Assets/Scripts/AsteroidCollisionHealthEffect.cs [hunk 2]
```diff
@@ -2,9 +2,15 @@
...
-    public override int Strength => damage;
-    public override HealthEffect Effect => HealthEffect.Damage;
+    public override int ComputeStrength(Collision collision = null) => collision == null
+        ? 0
+        : Mathf.RoundToInt(Mathf.Clamp(collision.impulse.magnitude / Time.fixedDeltaTime * (damageMultiplier / 1000),
+            1f,
+            maxDmg));
 
-    public int damage = 30;
+public override HealthEffect Effect => HealthEffect.Damage;
```
Comment (chuck-stein) [inline Assets/Scripts/AsteroidCollisionHealthEffect.cs]: indent this

File: Assets/Scripts/AsteroidCollisionHealthEffect.cs [hunk 3]
```diff
@@ -2,9 +2,15 @@
 
 public class AsteroidCollisionHealthEffect : HealthEffectBehavior
 {
-    public override int Strength => damage;
-    public override HealthEffect Effect => HealthEffect.Damage;
+    public override int ComputeStrength(Collision collision = null) => collision == null
+        ? 0
+        : Mathf.RoundToInt(Mathf.Clamp(collision.impulse.magnitude / Time.fixedDeltaTime * (damageMultiplier / 1000),
+            1f,
+            maxDmg));
```
Comment (chuck-stein) [inline Assets/Scripts/AsteroidCollisionHealthEffect.cs]: can you put this on the previous line?

File: Assets/Scripts/AsteroidCollisionHealthEffect.cs [hunk 4]
```diff
@@ -2,9 +2,15 @@
 
 public class AsteroidCollisionHealthEffect : HealthEffectBehavior
 {
-    public override int Strength => damage;
-    public override HealthEffect Effect => HealthEffect.Damage;
+    public override int ComputeStrength(Collision collision = null) => collision == null
+        ? 0
+        : Mathf.RoundToInt(Mathf.Clamp(collision.impulse.magnitude / Time.fixedDeltaTime * (damageMultiplier / 1000),
```
Comment (chuck-stein) [inline Assets/Scripts/AsteroidCollisionHealthEffect.cs]: I feel like this would read more intuitively if it was `collision.impulse.magnitude / Time.fixedDeltaTime / 1000f * damageMultiplier`

File: Assets/Prefabs/Asteroid.prefab [hunk 5]
```diff
@@ -236,7 +236,8 @@ MonoBehaviour:
   m_Script: {fileID: 11500000, guid: 5c0fbb3e4dda4edd8726a8e6438b0415, type: 3}
   m_Name: 
   m_EditorClassIdentifier: 
-  damage: 30
+  damageMultiplier: 1
```
Comment (chuck-stein) [inline Assets/Prefabs/Asteroid.prefab]: I played around with it a bit, can we make the `damageMultiplier` 1.5?

File: Assets/Scripts/GrappleGunBehavior.cs [hunk 6]
```diff
@@ -47,12 +49,31 @@ private bool GrappleTargetDestroyed()
...
 
+    private void BuildGrapple(RaycastHit hit)
+    {
+        joint = player.AddComponent<SpringJoint>();
+        joint.autoConfigureConnectedAnchor = false;
+        joint.anchor = Vector3.zero;
+        joint.connectedBody = hit.rigidbody;
+        if (grappledObjOffset.HasValue)
+        {
+            joint.connectedAnchor = grappledObjOffset.Value.normalized;
```
Comment (chuck-stein) [inline Assets/Scripts/GrappleGunBehavior.cs:60]: why are you normalizing the offset?

File: Assets/Scripts/GrappleGunBehavior.cs [hunk 7]
```diff
@@ -101,11 +124,13 @@ private void FixedUpdate()
     {
         ReticleEffect();
         if (!grappling) return;
-        if (GrappleBroken()) StopGrapple(); // TODO: add particle effects, SFX, and/or animation to indicate what happened
+        if (breakable && GrappleBroken()) StopGrapple(); // TODO: add particle effects, SFX, and/or animation to indicate what happened
         else
         {
+            joint.maxDistance *= 0.975f;
```
Comment (chuck-stein) [inline Assets/Scripts/GrappleGunBehavior.cs:130]: Shouldn't `maxDistance` just be set to the distance between the player and asteroid, so that it can never grow?

File: Assets/Scripts/GrappleGunBehavior.cs [hunk 8]
```diff
@@ -125,8 +150,11 @@ private bool GrappleBroken()
     private Vector3 ComputeTugForce()
     {
         var grappleDirection = GrapplePoint() - player.transform.position;
-        var tugStrength = grappleDirection.magnitude / maxGrappleLength * maxRetractionForce;
-        var tugForce = grappleDirection.normalized * tugStrength;
+        var distFromGrapplePoint = grappleDirection.magnitude;
+        
+        var directionMultiplier = Mathf.Abs(Vector3.Dot(playerRb.velocity.normalized, grappleDirection.normalized) - 1);
+        var tensionMultiplier = 35f * directionMultiplier * Mathf.Clamp(distFromGrapplePoint / maxRetractionForce, 0.5f, 1);
```
Comment (chuck-stein) [inline Assets/Scripts/GrappleGunBehavior.cs:156]: Can you walk me through the math for the `directionMultiplier` and `tensionMultiplier` and what the goal of each is / how its achieved?

File: Assets/Scripts/MedicalCanisterHealthEffect.cs [hunk 9]
```diff
@@ -1,6 +1,10 @@
-﻿public class MedicalCanisterHealthEffect : HealthEffectBehavior
+﻿using JetBrains.Annotations;
+using UnityEngine;
+
+public class MedicalCanisterHealthEffect : HealthEffectBehavior
 {
-    public override int Strength => healAmount;
+    public override int ComputeStrength([CanBeNull] Collision collision = null) => healAmount;
```
Comment (chuck-stein) [inline Assets/Scripts/MedicalCanisterHealthEffect.cs:6]: Can you include `[CanBeNull]` each place this is overriden as well?

File: Assets/Scripts/ProceduralGeneration.cs [hunk 10]
```diff
@@ -31,6 +31,9 @@ public class ProceduralGeneration : MonoBehaviour
 
     // Indices match up to values of EntityType enum
     public float[] spawnProbability;
+    [Tooltip("Will there be a black hole to chase the player")]
+    public bool chaseLevel = true;
```
Comment (chuck-stein) [inline Assets/Scripts/ProceduralGeneration.cs]: Can you revert the changes in this file to avoid merge conflicts? I did the same exact thing in my branch (except I called it `disableBlackHole`, because it's also useful for debugging)

**Review by chuck-stein — CHANGES_REQUESTED** [review]
some questions and small tweak requests

**Review by chuck-stein — CHANGES_REQUESTED** [review]
Fix exit portal yeeting you backwards into the black hole when grappled

File: Assets/Scripts/GrappleGunBehavior.cs (hunk 6 above)
Comment (Unix-Code) [reply Assets/Scripts/GrappleGunBehavior.cs:60]: For some weird reason, joints work in a strange way, where the connectedAnchor is proportional local coords where 0.5 is 0.5 of the object's size away from the center, so I normalize it to place the offset in the right spot.

File: Assets/Scripts/ProceduralGeneration.cs (hunk 10 above)
Comment (Unix-Code) [reply Assets/Scripts/ProceduralGeneration.cs]: Done

File: Assets/Prefabs/ExitPortal.prefab [hunk 11]
```diff
@@ -4826,6 +4826,7 @@ GameObject:
   - component: {fileID: 108504411327657124}
   - component: {fileID: 834488345}
   - component: {fileID: 834488344}
+  - component: {fileID: 5294384722759509742}
   m_Layer: 8
   m_Name: ExitPortal
```
Comment (chuck-stein) [inline Assets/Prefabs/ExitPortal.prefab:4831]: Why isn't the portal's rigidbody kinematic?

File: Assets/Prefabs/ExitPortal.prefab [hunk 12]
```diff
@@ -9715,6 +9716,22 @@ MonoBehaviour:
...
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_GameObject: {fileID: 1235273854538230}
+  serializedVersion: 2
+  m_Mass: 1000
+  m_Drag: 0
+  m_AngularDrag: 0
+  m_UseGravity: 0
+  m_IsKinematic: 0
```
Comment (chuck-stein) [inline Assets/Prefabs/ExitPortal.prefab:9731]: Why isn't the portal's rigidbody kinematic? Will it not work with the spring joint? If you can't make it kinematic, can you make the mass catastrophically large so nothing can ever move it?

File: Assets/Scenes/Prototype.unity (hunk 1 above)
Comment (chuck-stein) [reply Assets/Scenes/Prototype.unity:4408]: It looks like there are still changes from master. What did you change?

File: Assets/Scripts/AsteroidCollisionHealthEffect.cs [hunk 13]
```diff
@@ -1,16 +1,14 @@
...
 
 public class AsteroidCollisionHealthEffect : HealthEffectBehavior
 {
-    public override int ComputeStrength(Collision collision = null) => collision == null
+    public override int ComputeStrength([CanBeNull] Collision collision = null) => collision == null
         ? 0
-        : Mathf.RoundToInt(Mathf.Clamp(collision.impulse.magnitude / Time.fixedDeltaTime * (damageMultiplier / 1000),
-            1f,
-            maxDmg));
+        : Mathf.RoundToInt(Mathf.Clamp(collision.impulse.magnitude / Time.fixedDeltaTime / 1000 * damageMultiplier,1f, maxDmg));
```
Comment (chuck-stein) [inline Assets/Scripts/AsteroidCollisionHealthEffect.cs:8]: I meant move maxDmg to the same line as 1f but keep them on a separate line from the mathy one, but oh well

**Review by chuck-stein — APPROVED** [review]
good enough to merge but I left some comments for small things that would be nice to change if you can

File: Assets/Scripts/GrappleGunBehavior.cs (hunk 8 above)
Comment (chuck-stein) [reply Assets/Scripts/GrappleGunBehavior.cs:156]: ...pls?

File: Assets/Scripts/GrappleGunBehavior.cs (hunk 6 above)
Comment (chuck-stein) [reply Assets/Scripts/GrappleGunBehavior.cs:60]: I searched for what you were talking about but all I found is that the coordinates are in local space not world space. That doesn't mean 0.5 is 0.5 of the object's size away from the center. What I found says you should use `grappledTransform.InverseTransformPoint(hit.point)`. But if you don't want to do that I can do it when I merge in your changes.

File: Assets/Scenes/Prototype.unity (hunk 1 above)
Comment (Unix-Code) [reply Assets/Scenes/Prototype.unity:4408]: No clue?

Comment (chuck-stein) [reply Assets/Scenes/Prototype.unity:4408]: Then can you rollback to master to avoid merge conflicts?

File: Assets/Scripts/GrappleGunBehavior.cs (hunk 6 above)
Comment (Unix-Code) [reply Assets/Scripts/GrappleGunBehavior.cs:60]: I'm leaving this up to you

File: Assets/Scripts/GrappleGunBehavior.cs (hunk 8 above)
Comment (Unix-Code) [reply Assets/Scripts/GrappleGunBehavior.cs:156]: Basically, it applies the same boost the thruster does when the grapple is pulling you in an opposite direction of your current velocity. What it translates to is that it whips you backwards harder (more rigid rope).

File: Assets/Prefabs/ExitPortal.prefab (hunk 12 above)
Comment (Unix-Code) [reply Assets/Prefabs/ExitPortal.prefab:9731]: It has frozen constraints, so it can't be moved even on collision (if it's collider wasn't actually a trigger ofc). So, there's no concern of it every moving due to physics...

File: Assets/Scripts/AsteroidCollisionHealthEffect.cs (hunk 13 above)
Comment (Unix-Code) [reply Assets/Scripts/AsteroidCollisionHealthEffect.cs:8]: 🤷‍♂️

File: Assets/Prefabs/ExitPortal.prefab (hunk 12 above)
Comment (chuck-stein) [reply Assets/Prefabs/ExitPortal.prefab:9731]: oh ok perfect, I forgot about constraints
